File: ai-assistant/app/embeddings.py
"""
Turns text into a vector. Used both for indexing tasks (ingestion) and for
embedding the user's question at query time.
"""

from __future__ import annotations

from openai import AsyncOpenAI
from app.config import settings

_client = AsyncOpenAI(api_key=settings.openai_api_key)


async def embed_text(text: str) -> list[float]:
    # Truncate defensively — embedding models have a token limit, and a
    # runaway-long task description shouldn't crash ingestion.
    truncated = text[:8000]

    response = await _client.embeddings.create(
        model=settings.embedding_model,
        input=truncated,
    )
    return response.data[0].embedding

File: ai-assistant/src/embeddings.ts
/**
 * Turns text into a vector. Used for indexing tasks and embedding questions.
 */

import { OpenAIEmbeddings } from '@langchain/openai';
import { getConfig } from './config.js';

let embeddings: OpenAIEmbeddings | null = null;

function getEmbeddings(): OpenAIEmbeddings {
  if (embeddings) return embeddings;
  const cfg = getConfig();
  embeddings = new OpenAIEmbeddings({
    apiKey: cfg.openaiApiKey,
    model: cfg.embeddingModel,
    dimensions: cfg.embeddingDimensions,
  });
  return embeddings;
}

export async function embedText(text: string): Promise<number[]> {
  // Truncate defensively — embedding models have a token limit.
  const truncated = text.slice(0, 8000);
  const vector = await getEmbeddings().embedQuery(truncated);
  return vector;
}

/** Format as pgvector literal string for ::vector casts. */
export function embeddingLiteral(embedding: number[]): string {
  return `[${embedding.join(',')}]`;
}
